expression/builtin_other.go

```diff
@@ -320,11 +589,53 @@ func (b *builtinInDurationSig) evalInt(row chunk.Row) (int64, bool, error) {
 // builtinInJSONSig see https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_in
 type builtinInJSONSig struct {
 	baseBuiltinFunc
+	nonConstArgs []Expression
+	hashSet      map[string]bool
+	threshold    int
+}
+
+func (b *builtinInJSONSig) buildHashMapForConstArgs(ctx sessionctx.Context) error {
+	b.nonConstArgs = make([]Expression, 0, len(b.args))
+	b.nonConstArgs = append(b.nonConstArgs, b.args[0])
+	b.hashSet = make(map[string]bool, len(b.args)-1)
+	count := 0
+	for i := 1; i < len(b.args); i++ {
+		if b.args[i].ConstItem() {
+			val, isNull, err := b.args[i].EvalJSON(ctx, chunk.Row{})
+			if err != nil {
+				return err
+			}
+			if isNull {
+				b.nonConstArgs = append(b.nonConstArgs, b.args[i])
+				continue
+			}
+			json, err := val.MarshalJSON()
```

Comment:
For performance, it's better to implement BinaryJSON.ToHashKey() and use `BinaryJSON.TypeCode` and `BinaryJSON.Value` and hash key.